### PR TITLE
Implement the unicode-bidi CSS property

### PR DESCRIPTION
**Pull request by mbrubeck**
r? @pcwalton

[Review on Reviewable](https://reviewable.io/reviews/servo/servo/6721)

### CODE REVIEW

File: components/gfx/text/glyph.rs [hunk 1]
```diff
@@ -782,17 +784,23 @@ impl<'a> Iterator for GlyphIterator<'a> {
             self.next_glyph_range()
```
Comment (pcwalton) [inline components/gfx/text/glyph.rs]: nit: While you're here, could you change this into an early return? Then you can de-indent all your code below.

File: components/gfx/text/glyph.rs [hunk 2]
```diff
@@ -782,17 +784,23 @@ impl<'a> Iterator for GlyphIterator<'a> {
...
+            self.char_index = self.char_index + if self.store.is_rtl {
+                CharIndex(-1)
+            } else {
+                CharIndex(1)
+            };
+            let i = self.char_index;
+            if !self.char_range.contains(i) {
+                return None
+            }
+            assert!(i < self.store.char_len());
```
Comment (pcwalton) [inline components/gfx/text/glyph.rs]: nit: `debug_assert!` — I bet this code is performance critical.

**Comment by mbrubeck** [comment]
Review comments addressed. r? @pcwalton

**Comment by metajack** [comment]
pinging @pcwalton

**Comment by jdm** [comment]
```


/css21_dev/html4/inline-formatting-context-007.htm
--------------------------------------------------
FAIL [Parent]
```
Can you take a look and see if anything in that test should be affected by this?

**Comment by mbrubeck** [comment]
Also:
```
/html/semantics/text-level-semantics/the-bdi-element/bdi-neutral-nested.html
----------------------------------------------------------------------------
PASS expected FAIL [Parent]
```

Both of these unexpected results are Mac-only.  (The Linux builders are green.)

* `inline-formatting-context-007.htm` was previously failing on both platforms, now failing on Mac only.
* `bdi-neutral-nested.html` was previously failing on both platforms, not failing on Linux only.

Looking at the reftest analyzer, even on the failing platforms, the rendering with the patch is an improvement over the rendering before.  The remaining failures appear to be unrelated bugs.  For example `inline-formatting-context-007.htm` has extra top/bottom borders, which are also the cause of the previously-existing failures in `inline-formatting-context-004.htm` and `inline-formatting-context-005.htm`.

I'll retry this with adjusted text expectations.

**Pull request merged**